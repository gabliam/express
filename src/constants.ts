/** Type in registry */
export const TYPE = {
  Controller: 'ControllerType'
};

export const METADATA_KEY = {
  controller: '_controller',
  controllerMethod: '_controller-method',
  controllerParameter: '_controller-parameter',
  middleware: '_middleware',
  MiddlewareConfig: '_middlewareConfig',
  MiddlewareErrorConfig: '_middlewareErrorConfig'
};

export enum PARAMETER_TYPE {
  REQUEST,
  RESPONSE,
  PARAMS,
  QUERY,
  BODY,
  HEADERS,
  COOKIES,
  NEXT
}

export const DEFAULT_ROUTING_ROOT_PATH = '/';

export const EXPRESS_PLUGIN_CONFIG = Symbol('GABLIAM/EXPRESS_PLUGIN_CONFIG');

export const APP = Symbol('GABLIAM/EXPRESS_APP');
export const SERVER = Symbol('GABLIAM/EXPRESS_SERVER');
