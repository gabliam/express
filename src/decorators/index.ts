export * from './controller';
export * from './middleware';
export * from './config';
export * from './method';
export * from './params';
