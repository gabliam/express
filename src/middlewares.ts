import { ExpressConfig } from './interfaces';

export class MiddlewareConfig {
  private _middlewares: ExpressConfig[] = [];

  private _errorMiddlewares: ExpressConfig[] = [];


  addMiddleware(middleware: ExpressConfig) {
    this._middlewares.push(middleware);
  }

  addErrorMiddleware(middleware: ExpressConfig) {
    this._errorMiddlewares.push(middleware);
  }

  get middlewares() {
    return this._middlewares.slice();
  }

  get errorMiddlewares() {
    return this._errorMiddlewares.slice();
  }
}
