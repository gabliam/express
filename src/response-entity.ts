import { OutgoingHttpHeaders } from 'http';
import * as HttpStatus from 'http-status-codes';

export class ResponseEntity<T = any> {
  private _status = HttpStatus.OK;

  private _headers: OutgoingHttpHeaders = {};

  private _body: T | null = null;

  constructor(body: T, status: number = 200) {
    this._body = body;
    this._status = status;
  }

  get status() {
    return this._status;
  }

  get headers() {
    return this._headers;
  }

  get body() {
    return this._body;
  }

  hasHeader() {
    return Object.keys(this._headers).length !== 0;
  }

  addHeader(headerName: string, value: any) {
    this._headers[headerName] = value;
  }
}

/**
 * A shortcut for creating a ResponseEntity with the given body and the status set to OK.
 * @param body
 */
export function ok<T = any>(body?: T) {
  return new ResponseEntity(body, HttpStatus.OK);
}

/**
 * Create a ResponseEntity with an ACCEPTED status.
 */
export function accepted() {
  return new ResponseEntity(null, HttpStatus.ACCEPTED);
}

/**
 * Create a ResponseEntity with a BAD_REQUEST status.
 */
export function badRequest() {
  return new ResponseEntity(null, HttpStatus.BAD_REQUEST);
}

/**
 * Create a ResponseEntity with a NO_CONTENT status.
 */
export function noContent() {
  return new ResponseEntity(null, HttpStatus.NO_CONTENT);
}

/**
 * Create a ResponseEntity with a NOT_FOUND status.
 */
export function notFound() {
  return new ResponseEntity(null, HttpStatus.NOT_FOUND);
}
