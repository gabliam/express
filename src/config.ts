import { PluginConfig, Bean, Value } from '@gabliam/core';
import { DEFAULT_ROUTING_ROOT_PATH, EXPRESS_PLUGIN_CONFIG } from './constants';
import { ExpressPluginConfig } from './interfaces';
import * as Joi from 'joi';

@PluginConfig()
export class RestPluginConfig {
    @Value('application.express.rootPath', Joi.string())
    rootPath = DEFAULT_ROUTING_ROOT_PATH;

    @Value('application.express.port', Joi.number().positive())
    port: number = process.env.PORT || 3000;

    @Value('application.express.hostname', Joi.string())
    hostname: string;


    @Bean(EXPRESS_PLUGIN_CONFIG)
    restConfig(): ExpressPluginConfig {
        return {
            rootPath: this.rootPath,
            port: this.port,
            hostname: this.hostname
        };
    }
}
