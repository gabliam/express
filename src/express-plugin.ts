import {
  interfaces as coreInterfaces,
  inversifyInterfaces,
  Scan,
  Registry
} from '@gabliam/core';
import {
  TYPE,
  METADATA_KEY,
  EXPRESS_PLUGIN_CONFIG,
  APP,
  SERVER,
  PARAMETER_TYPE
} from './constants';
import { getMiddlewares } from './metadata';
import { cleanPath } from './utils';
import {
  ExpressConfigMetadata,
  ExpressPluginConfig,
  ControllerMetadata,
  ControllerMethodMetadata,
  ParameterMetadata,
  ControllerParameterMetadata
} from './interfaces';
import * as express from 'express';
import * as d from 'debug';
import * as http from 'http';
import { MiddlewareConfig } from './middlewares';

const debug = d('Gabliam:Plugin:ExpressPlugin');

@Scan(__dirname)
export class ExpressPlugin implements coreInterfaces.GabliamPlugin {
  /**
   * binding phase
   *
   * Bind all controller and bind express app
   * @param  {inversifyInterfaces.Container} container
   * @param  {Registry} registry
   */
  bind(container: inversifyInterfaces.Container, registry: Registry) {
    container.bind(APP).toConstantValue(express());
    registry
      .get(TYPE.Controller)
      .forEach(({ id, target }) =>
        container.bind<any>(id).to(target).inSingletonScope()
      );

    container.bind(MiddlewareConfig).toConstantValue(new MiddlewareConfig());
  }

  build(container: inversifyInterfaces.Container, registry: Registry) {
    this.buildExpressConfig(container, registry);
    this.buildControllers(container, registry);
    this.buildExpressErrorConfig(container, registry);
  }

  /**
   * Management of @middleware decorator in config class
   *
   * @param  {inversifyInterfaces.Container} container
   * @param  {Registry} registry
   * @param  {any} confInstance
   */
  config(
    container: inversifyInterfaces.Container,
    registry: Registry,
    confInstance: any
  ) {
    const middlewareConfig = container.get<MiddlewareConfig>(MiddlewareConfig);
    // if config class has a @middleware decorator, add in this.middlewares for add it in building phase
    if (
      Reflect.hasMetadata(
        METADATA_KEY.MiddlewareConfig,
        confInstance.constructor
      )
    ) {
      const metadataList: ExpressConfigMetadata[] = Reflect.getOwnMetadata(
        METADATA_KEY.MiddlewareConfig,
        confInstance.constructor
      );

      metadataList.forEach(({ key, order }) => {
        middlewareConfig.addMiddleware({
          order,
          instance: confInstance[key].bind(confInstance[key])
        });
      });
    }

    // if config class has a @middleware decorator, add in this.errorMiddlewares for add it in building phase
    if (
      Reflect.hasMetadata(
        METADATA_KEY.MiddlewareErrorConfig,
        confInstance.constructor
      )
    ) {
      const metadataList: ExpressConfigMetadata[] = Reflect.getOwnMetadata(
        METADATA_KEY.MiddlewareErrorConfig,
        confInstance.constructor
      );

      metadataList.forEach(({ key, order }) => {
        middlewareConfig.addErrorMiddleware({
          order,
          instance: confInstance[key].bind(confInstance[key])
        });
      });
    }
  }

  /**
   * Build express middleware
   *
   * @param  {inversifyInterfaces.Container} container
   * @param  {Registry} registry
   */
  private buildExpressConfig(
    container: inversifyInterfaces.Container,
    registry: Registry
  ) {
    const middlewareConfig = container.get<MiddlewareConfig>(MiddlewareConfig);
    const app = container.get<express.Application>(APP);
    middlewareConfig.middlewares
      .sort((a, b) => a.order - b.order)
      .forEach(({ instance }) => instance(app));
  }

  /**
   * Build express error middleware
   *
   * @param  {inversifyInterfaces.Container} container
   * @param  {Registry} registry
   */
  private buildExpressErrorConfig(
    container: inversifyInterfaces.Container,
    registry: Registry
  ) {
    const middlewareConfig = container.get<MiddlewareConfig>(MiddlewareConfig);
    const app = container.get<express.Application>(APP);
    middlewareConfig.errorMiddlewares
      .sort((a, b) => a.order - b.order)
      .forEach(({ instance }) => instance(app));
  }

  private buildControllers(
    container: inversifyInterfaces.Container,
    registry: Registry
  ) {
    const restConfig = container.get<ExpressPluginConfig>(
      EXPRESS_PLUGIN_CONFIG
    );

    debug('registerControllers', TYPE.Controller);
    const controllerIds = registry.get(TYPE.Controller);
    controllerIds.forEach(({ id: controllerId }) => {
      const controller = container.get<object>(controllerId);

      const controllerMetadata: ControllerMetadata = Reflect.getOwnMetadata(
        METADATA_KEY.controller,
        controller.constructor
      );

      const controllerMiddlewares = getMiddlewares(
        container,
        controller.constructor
      );

      const methodMetadatas: ControllerMethodMetadata[] = Reflect.getOwnMetadata(
        METADATA_KEY.controllerMethod,
        controller.constructor
      );

      const parameterMetadata: ControllerParameterMetadata = Reflect.getOwnMetadata(
        METADATA_KEY.controllerParameter,
        controller.constructor
      );

      if (controllerMetadata && methodMetadatas) {
        const router = express.Router();
        const routerPath = cleanPath(
          `${restConfig.rootPath}${controllerMetadata.path}`
        );

        debug(`New route : "${routerPath}"`);

        methodMetadatas.forEach((methodMetadata: ControllerMethodMetadata) => {
          let paramList: ParameterMetadata[] = [];
          if (parameterMetadata) {
            paramList = parameterMetadata[methodMetadata.key] || [];
          }
          const methodMetadataPath = cleanPath(methodMetadata.path);
          const methodMiddlewares = getMiddlewares(
            container,
            controller.constructor,
            methodMetadata.key
          );
          debug(methodMetadataPath);
          debug({ methodMiddlewares, controllerMiddlewares });
          const handler: express.RequestHandler = this.handlerFactory(
            container,
            controllerId,
            methodMetadata.key,
            controllerMetadata.json,
            paramList
          );
          (router as any)[methodMetadata.method](
            methodMetadataPath,
            ...controllerMiddlewares,
            ...methodMiddlewares,
            handler
          );
        });
        const app = container.get<express.Application>(APP);
        app.use(routerPath, router);
      }
    });
  }

  private handlerFactory(
    container: inversifyInterfaces.Container,
    controllerId: any,
    key: string,
    json: boolean,
    parameterMetadata: ParameterMetadata[]
  ): express.RequestHandler {
    return (
      req: express.Request,
      res: express.Response,
      next: express.NextFunction
    ) => {
      const controller = container.get<any>(controllerId);
      const args = this.extractParameters(
        controller,
        key,
        req,
        res,
        next,
        parameterMetadata
      );
      const result: any = controller[key](...args);

      // try to resolve promise
      if (result && result instanceof Promise) {
        result
          .then((value: any) => {
            if (value !== undefined && !res.headersSent) {
              if (json) {
                res.json(value);
              } else {
                res.send(value);
              }
            }
          })
          .catch((error: any) => {
            next(error);
          });
      } else if (result !== undefined && !res.headersSent) {
        if (json) {
          res.json(result);
        } else {
          res.send(result);
        }
      }
    };
  }

  async destroy(container: inversifyInterfaces.Container, registry: Registry) {
    await this.stop(container, registry);
  }

  async stop(container: inversifyInterfaces.Container, registry: Registry) {
    try {
      // server can be undefined (if start is not called)
      const server = container.get<http.Server>(SERVER);
      return new Promise<void>(resolve => {
        server.close(() => resolve());
      });
    } catch (e) {}
  }

  async start(container: inversifyInterfaces.Container, registry: Registry) {
    const restConfig = container.get<ExpressPluginConfig>(
      EXPRESS_PLUGIN_CONFIG
    );
    const app = container.get<express.Application>(APP);
    const port = restConfig.port;
    app.set('port', port);

    const server = http.createServer(app);
    server.listen(port, restConfig.hostname);
    server.on('error', onError);
    server.on('listening', onListening);
    container.bind(SERVER).toConstantValue(server);

    function onError(error: NodeJS.ErrnoException): void {
      // tslint:disable-next-line:curly
      if (error.syscall !== 'listen') throw error;
      const bind = typeof port === 'string' ? 'Pipe ' + port : 'Port ' + port;
      switch (error.code) {
        case 'EACCES':
          console.error(`${bind} requires elevated privileges`);
          process.exit(1);
          break;
        case 'EADDRINUSE':
          console.error(`${bind} is already in use`);
          process.exit(1);
          break;
        default:
          throw error;
      }
    }

    function onListening(): void {
      const addr = server.address();
      const bind =
        typeof addr === 'string' ? `pipe ${addr}` : `port ${addr.port}`;
      console.log(`Listening on ${bind}`);
    }
  }

  private extractParameters(
    target: any,
    key: string,
    req: express.Request,
    res: express.Response,
    next: express.NextFunction,
    params: ParameterMetadata[]
  ): any[] {
    const args = [];
    if (!params || !params.length) {
      return [req, res, next];
    }
    const getParam = this.getFuncParam(target, key);

    for (const item of params) {
      switch (item.type) {
        default:
          args[item.index] = res;
          break; // response
        case PARAMETER_TYPE.REQUEST:
          args[item.index] = getParam(req, null, item);
          break;
        case PARAMETER_TYPE.NEXT:
          args[item.index] = next;
          break;
        case PARAMETER_TYPE.PARAMS:
          args[item.index] = getParam(req, 'params', item);
          break;
        case PARAMETER_TYPE.QUERY:
          args[item.index] = getParam(req, 'query', item);
          break;
        case PARAMETER_TYPE.BODY:
          args[item.index] = getParam(req, 'body', item);
          break;
        case PARAMETER_TYPE.HEADERS:
          args[item.index] = getParam(req, 'headers', item);
          break;
        case PARAMETER_TYPE.COOKIES:
          args[item.index] = getParam(req, 'cookies', item);
          break;
      }
    }
    args.push(req, res, next);
    return args;
  }

  private getFuncParam(target: any, key: string) {
    return (
      source: any,
      paramType: string | null,
      itemParam: ParameterMetadata
    ) => {
      const name = itemParam.parameterName;

      let param = source;
      if (paramType !== null && source[paramType]) {
        param = source[paramType];
      }

      let res = param[name];
      if (res) {
        /**
         * For query, all value sare considered to string value.
         * If the query waits for a Number, we try to convert the value
         */
        if (paramType === 'query' || paramType === 'params') {
          const type: Function[] = Reflect.getMetadata(
            'design:paramtypes',
            target,
            key
          );
          if (Array.isArray(type) && type[itemParam.index]) {
            try {
              if (type[itemParam.index].name === 'Number') {
                // parseFloat for compatibility with integer and float
                res = Number.parseFloat(res);
              }
            } catch (e) {}
          }
        }
        return res;
      } else {
        return paramType === 'query' ? undefined : param;
      }
    };
  }
}
