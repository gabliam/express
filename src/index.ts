import { APP, SERVER } from './constants';
import * as interfaces from './interfaces';
import { MiddlewareConfig } from './middlewares';
import { ExpressPlugin } from './express-plugin';

export * from './decorators';
export * from './response-entity';
export * from 'http-status-codes';
export { interfaces, APP, SERVER, MiddlewareConfig, ExpressPlugin as default };
