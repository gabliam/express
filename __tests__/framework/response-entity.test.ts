import { Controller, RestController, Get } from '../../src/index';
import * as e from 'express';
import { ExpressPluginTest } from '../express-plugin-test';
import * as supertest from 'supertest';
import {
  ResponseEntity,
  ok,
  accepted,
  badRequest,
  noContent,
  notFound
} from '../../src/response-entity';
import * as HttpStatus from 'http-status-codes';

let appTest: ExpressPluginTest;

beforeEach(async () => {
  appTest = new ExpressPluginTest();
});

afterEach(async () => {
  await appTest.destroy();
});

describe('Response entity Tests:', () => {
  [Controller, RestController].forEach(decorator => {
    describe(`decorator ${decorator.name}`, () => {
      test('should work for responseEntitie controller methods', async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          getTest(req: e.Request, res: e.Response) {
            return new ResponseEntity({ get: 'GET' });
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const response = await supertest(appTest.app)
          .get('/')
          .expect(HttpStatus.OK);
        expect(response).toMatchSnapshot();
      });

      test('should work for responseEntitie and custom Header controller methods', async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          getTest(req: e.Request, res: e.Response) {
            const resp = new ResponseEntity({ get: 'GET' });
            resp.addHeader('X-Test', 'tests');
            return resp;
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const response = await supertest(appTest.app)
          .get('/')
          .expect(HttpStatus.OK);
        expect(response).toMatchSnapshot();
      });

      test('should work for async responseEntitie controller methods', async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          async getTest(req: e.Request, res: e.Response) {
            return new Promise(resolve => {
              setTimeout(resolve, 100, new ResponseEntity({ get: 'GET' }));
            });
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const response = await supertest(appTest.app)
          .get('/')
          .expect(HttpStatus.OK);
        expect(response).toMatchSnapshot();
      });

      test('responseEntitie ok()', async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          async getTest(req: e.Request, res: e.Response) {
            return ok();
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const response = await supertest(appTest.app)
          .get('/')
          .expect(HttpStatus.OK);
        expect(response).toMatchSnapshot();
      });

      test(`responseEntitie ok({get: 'get'})`, async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          async getTest(req: e.Request, res: e.Response) {
            return ok({ get: 'get' });
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const response = await supertest(appTest.app)
          .get('/')
          .expect(HttpStatus.OK);
        expect(response).toMatchSnapshot();
      });

      test(`responseEntitie accepted()`, async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          async getTest(req: e.Request, res: e.Response) {
            return accepted();
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const response = await supertest(appTest.app)
          .get('/')
          .expect(HttpStatus.ACCEPTED);
        expect(response).toMatchSnapshot();
      });

      test(`responseEntitie badRequest()`, async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          async getTest(req: e.Request, res: e.Response) {
            return badRequest();
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const response = await supertest(appTest.app)
          .get('/')
          .expect(HttpStatus.BAD_REQUEST);
        expect(response).toMatchSnapshot();
      });

      test(`responseEntitie noContent()`, async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          async getTest(req: e.Request, res: e.Response) {
            return noContent();
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const response = await supertest(appTest.app)
          .get('/')
          .expect(HttpStatus.NO_CONTENT);
        expect(response).toMatchSnapshot();
      });

      test(`responseEntitie notFound()`, async () => {
        @decorator('/')
        class TestController {
          @Get('/')
          async getTest(req: e.Request, res: e.Response) {
            return notFound();
          }
        }
        appTest.addClass(TestController);
        await appTest.build();
        const response = await supertest(appTest.app)
          .get('/')
          .expect(HttpStatus.NOT_FOUND);
        expect(response).toMatchSnapshot();
      });
    });
  });
});
