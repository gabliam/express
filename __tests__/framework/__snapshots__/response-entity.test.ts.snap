// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Response entity Tests: decorator Controller responseEntitie accepted() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"4-K+iMpCQsduglOsYkdIUQZQMtaDM\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 202,
  "text": "null",
}
`;

exports[`Response entity Tests: decorator Controller responseEntitie badRequest() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"4-K+iMpCQsduglOsYkdIUQZQMtaDM\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 400,
  "text": "null",
}
`;

exports[`Response entity Tests: decorator Controller responseEntitie noContent() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "etag": "W/\\"4-K+iMpCQsduglOsYkdIUQZQMtaDM\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 204,
  "text": "",
}
`;

exports[`Response entity Tests: decorator Controller responseEntitie notFound() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"4-K+iMpCQsduglOsYkdIUQZQMtaDM\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 404,
  "text": "null",
}
`;

exports[`Response entity Tests: decorator Controller responseEntitie ok() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "0",
    "content-type": "application/json",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "",
}
`;

exports[`Response entity Tests: decorator Controller responseEntitie ok({get: 'get'}) 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"d-Lc+tjcJImP8Sf1HWyafiv3+lfrc\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "{\\"get\\":\\"get\\"}",
}
`;

exports[`Response entity Tests: decorator Controller should work for async responseEntitie controller methods 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"d-navs5TfboJktcSL9TcDVd7fN4Zk\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "{\\"get\\":\\"GET\\"}",
}
`;

exports[`Response entity Tests: decorator Controller should work for responseEntitie and custom Header controller methods 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"d-navs5TfboJktcSL9TcDVd7fN4Zk\\"",
    "x-powered-by": "Express",
    "x-test": "tests",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "{\\"get\\":\\"GET\\"}",
}
`;

exports[`Response entity Tests: decorator Controller should work for responseEntitie controller methods 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"d-navs5TfboJktcSL9TcDVd7fN4Zk\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "{\\"get\\":\\"GET\\"}",
}
`;

exports[`Response entity Tests: decorator RestController responseEntitie accepted() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"4-K+iMpCQsduglOsYkdIUQZQMtaDM\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 202,
  "text": "null",
}
`;

exports[`Response entity Tests: decorator RestController responseEntitie badRequest() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"4-K+iMpCQsduglOsYkdIUQZQMtaDM\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 400,
  "text": "null",
}
`;

exports[`Response entity Tests: decorator RestController responseEntitie noContent() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "etag": "W/\\"4-K+iMpCQsduglOsYkdIUQZQMtaDM\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 204,
  "text": "",
}
`;

exports[`Response entity Tests: decorator RestController responseEntitie notFound() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"4-K+iMpCQsduglOsYkdIUQZQMtaDM\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 404,
  "text": "null",
}
`;

exports[`Response entity Tests: decorator RestController responseEntitie ok() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "0",
    "content-type": "application/json",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "",
}
`;

exports[`Response entity Tests: decorator RestController responseEntitie ok({get: 'get'}) 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"d-Lc+tjcJImP8Sf1HWyafiv3+lfrc\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "{\\"get\\":\\"get\\"}",
}
`;

exports[`Response entity Tests: decorator RestController should work for async responseEntitie controller methods 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"d-navs5TfboJktcSL9TcDVd7fN4Zk\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "{\\"get\\":\\"GET\\"}",
}
`;

exports[`Response entity Tests: decorator RestController should work for responseEntitie and custom Header controller methods 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"d-navs5TfboJktcSL9TcDVd7fN4Zk\\"",
    "x-powered-by": "Express",
    "x-test": "tests",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "{\\"get\\":\\"GET\\"}",
}
`;

exports[`Response entity Tests: decorator RestController should work for responseEntitie controller methods 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"d-navs5TfboJktcSL9TcDVd7fN4Zk\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "{\\"get\\":\\"GET\\"}",
}
`;
